components/tidb_query/src/codec/collation/mod.rs

```diff
@@ -91,7 +93,128 @@ impl Collator for CollatorBinary {
     }
 
     #[inline]
-    fn sort_hash<H: Hasher>(_state: &mut H, _bstr: &[u8]) -> Result<()> {
-        unimplemented!()
+    fn sort_hash<H: Hasher>(state: &mut H, bstr: &[u8]) -> Result<()> {
+        bstr.hash(state);
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct SortKey<T, C: Collator>
+where
+    T: AsRef<[u8]>,
+{
+    inner: T,
+    _phantom: PhantomData<C>,
+}
+
+impl<T, C: Collator> SortKey<T, C>
+where
+    T: AsRef<[u8]>,
+{
+    #[inline]
+    pub fn new(inner: T) -> Result<Self> {
+        C::validate(inner.as_ref())?;
+        Ok(Self {
+            inner,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Create SortKey from unchecked bytes.
+    ///
+    /// # Safety
+    ///
+    /// The `Ord`, `Hash`, `PartialEq` and more implementations assume that the bytes are
+    /// valid for the certain collator. The violation will cause panic.
+    #[inline]
+    pub unsafe fn new_unchecked(inner: T) -> Self {
```

Comment:
Panic is still safe. Does it really need to be `unsafe`?

Reply:
I think `NotNan` is an example of this. https://docs.rs/ordered-float/1.0.2/ordered_float/struct.NotNan.html#method.unchecked_new

Reply:
Well, it's not necessary now. But we can remove the validate in `sort_hash` later.

Reply:
Humm. You're right, I'll remove the unsafe.

Reply:
prefer unsafe, i think a better example is https://doc.rust-lang.org/std/str/fn.from_utf8_unchecked.html

unsafe means the memory may not be interpreted correctly with current type.

And in sort_hash, maybe we can use assert is_ok instead of unwrap ( it’s not necessary)